pkg/lib/authn/sso/adfs.go

```diff
@@ -0,0 +1,128 @@
+package sso
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/clock"
+	"github.com/authgear/authgear-server/pkg/util/validation"
+)
+
+type ADFSImpl struct {
+	Clock                    clock.Clock
+	RedirectURL              RedirectURLProvider
+	ProviderConfig           config.OAuthSSOProviderConfig
+	Credentials              config.OAuthClientCredentialsItem
+	LoginIDNormalizerFactory LoginIDNormalizerFactory
+}
+
+func (*ADFSImpl) Type() config.OAuthSSOProviderType {
+	return config.OAuthSSOProviderTypeADFS
+}
+
+func (f *ADFSImpl) Config() config.OAuthSSOProviderConfig {
+	return f.ProviderConfig
+}
+
+func (f *ADFSImpl) getOpenIDConfiguration() (*OIDCDiscoveryDocument, error) {
+	endpoint := f.ProviderConfig.DiscoveryDocumentEndpoint
+	return FetchOIDCDiscoveryDocument(http.DefaultClient, endpoint)
+}
+
+func (f *ADFSImpl) GetAuthURL(param GetAuthURLParam) (string, error) {
+	c, err := f.getOpenIDConfiguration()
+	if err != nil {
+		return "", err
+	}
+	return c.MakeOAuthURL(OIDCAuthParams{
+		ProviderConfig: f.ProviderConfig,
+		RedirectURI:    f.RedirectURL.SSOCallbackURL(f.ProviderConfig).String(),
+		Nonce:          param.Nonce,
+		State:          param.State,
+	}), nil
+}
+
+func (f *ADFSImpl) GetAuthInfo(r OAuthAuthorizationResponse, param GetAuthInfoParam) (authInfo AuthInfo, err error) {
+	return f.OpenIDConnectGetAuthInfo(r, param)
+}
+
+func (f *ADFSImpl) OpenIDConnectGetAuthInfo(r OAuthAuthorizationResponse, param GetAuthInfoParam) (authInfo AuthInfo, err error) {
+	c, err := f.getOpenIDConfiguration()
+	if err != nil {
+		return
+	}
+
+	// OPTIMIZE(sso): Cache JWKs
+	keySet, err := c.FetchJWKs(http.DefaultClient)
+	if err != nil {
+		return
+	}
+
+	var tokenResp AccessTokenResp
+	jwtToken, err := c.ExchangeCode(
+		http.DefaultClient,
+		f.Clock,
+		r.Code,
+		keySet,
+		f.ProviderConfig.ClientID,
+		f.Credentials.ClientSecret,
+		f.RedirectURL.SSOCallbackURL(f.ProviderConfig).String(),
+		param.Nonce,
+		&tokenResp,
+	)
+	if err != nil {
+		return
+	}
+
+	claims, err := jwtToken.AsMap(context.TODO())
+	if err != nil {
+		return
+	}
+
+	sub, ok := claims["sub"].(string)
+	if !ok {
+		err = OAuthProtocolError.New("sub not found in ID token")
+		return
+	}
+
+	// The upn claim is documented here.
+	// https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/operations/configuring-alternate-login-id
+	upn, ok := claims["upn"].(string)
+	if !ok {
+		err = OAuthProtocolError.New("upn not found in ID token")
+		return
+	}
+
+	var email string
+	var preferredUsername string
+
+	err = validation.FormatEmail{}.CheckFormat(upn)
+	if err == nil {
+		// upn looks like an email address.
+		normalizer := f.LoginIDNormalizerFactory.NormalizerWithLoginIDType(config.LoginIDKeyTypeEmail)
+		email, err = normalizer.Normalize(upn)
+		if err != nil {
+			return
+		}
+	} else {
+		// upn does NOT look like an email address.
+		preferredUsername = upn
```

Comment:
Or we should always fill `upn` in `preferredUsername`? So adfs always have `preferredUsername` consistently. I am open to this

Reply:
Let me change to always fill in upn.